File: backend/feature/post/src/domain/entity/post_info.rs
use chrono::{DateTime, Utc};
use label::domain::entity::label::Label;
use regex::Regex;

use crate::domain::error::post_error::PostError;

pub struct PostInfo {
    pub id: i32,
    pub semantic_id: String,
    pub title: String,
    pub description: String,
    pub preview_image_url: Option<String>,
    pub labels: Vec<Label>,
    pub published_time: Option<DateTime<Utc>>,
}

impl PostInfo {
    pub fn validate_semantic_id(semantic_id: &str) -> Result<(), PostError> {
        if semantic_id.parse::<i32>().is_ok() {
            return Err(PostError::InvalidSemanticId);
        }

        let re = Regex::new(r"^[0-9a-zA-Z_-]+$").unwrap();
        if !re.is_match(semantic_id) {
            return Err(PostError::InvalidSemanticId);
        }

        Ok(())
    }
}

File: frontend/src/lib/post/domain/entity/postInfo.ts
import type { Label } from '$lib/label/domain/entity/label';

export class PostInfo {
	readonly id: number;
	readonly semanticId: string;
	readonly title: string;
	readonly description: string;
	readonly previewImageUrl: string | null;
	readonly previewImageMimeType: string | null;
	readonly previewImageSize: number | null;
	readonly labels: readonly Label[];
	readonly publishedTime: Date | null;
	readonly updatedTime: Date | null;

	constructor(props: {
		id: number;
		semanticId: string;
		title: string;
		description: string;
		previewImageUrl: string | null;
		previewImageMimeType: string | null;
		previewImageSize: number | null;
		labels: readonly Label[];
		publishedTime: Date | null;
		updatedTime: Date | null;
	}) {
		this.id = props.id;
		this.semanticId = props.semanticId;
		this.title = props.title;
		this.description = props.description;
		this.previewImageUrl = props.previewImageUrl;
		this.previewImageMimeType = props.previewImageMimeType;
		this.previewImageSize = props.previewImageSize;
		this.labels = props.labels;
		this.publishedTime = props.publishedTime;
		this.updatedTime = props.updatedTime;
	}
}
